Update syncs

lkatalin opened the pull request:
Related to #893 

#### Summary
- Update main to preprod sync to run only on changes to timestamp/snapshot and to sync to both preprod and prod.
- Update ceremony to main sync to also sync to preprod.

#### Release Note
NONE

#### Documentation
Updated `ORCHESTRATION.md`

Hayden-IO commented on .github/workflows/sync-main-to-prod.yml:
```diff
@@ -13,16 +13,25 @@
-      # When any repository metadata is changed.
-      - 'repository/repository/**'
+      # When timestamp or snapshot files are changed
+      - 'repository/repository/timestamp.json'
```
We'll need to check in another way, since this would also trigger when the timestamp is updated along with the root during a root signing event.

Hayden-IO replied:
We could add a step to diff what's in main vs what's in the bucket?

lkatalin replied:
You're totally right. I will push a new version soon.

lkatalin replied:
Updated

kommendorkapten commented on .github/workflows/sync-ceremony-to-main.yml:
```diff
@@ -13,12 +13,11 @@
-# thereby automatically triggering the `sync-main-to-preprod.yml` workflow.
+# This workflow is triggered when a ceremony branch is completed, and will:
+# - Create a PR that merges the completed ceremony branch to main
+# - Sync the main branch to the GCS preprod bucket
```
Technically it's not `main` that's synced to preprod bucket, right? It's the ceremony branch that's about to be merged to main. So something like `Sync the ceremony branch to the GCS preprod bucket` would be better I think.

lkatalin replied:
Updated

Hayden-IO commented on .github/workflows/sync-main-to-prod.yml:
```diff
@@ -13,16 +13,26 @@
+# - timestamp.json
+# - snapshot.json
+# - [0-9]+.snapshot.json
+# Under this condition, the entire main branch (all files) will sync
```
nit, the repository/repository folder on main branch, rather than all files

lkatalin replied:
Updated

Hayden-IO commented on .github/workflows/sync-main-to-prod.yml:
```diff
@@ -66,6 +76,19 @@ jobs:
+          # Note: other non-timestamp/snapshot changes should only occur during a ceremony, and
+          # will go through the sync-ceremony-* flow that hits main and preprod. This means there should never 
+          # be changed files in prod that have not also hit the main branch.
+          diff -qr repository/repository sigstore-tuf-root | grep -Po '([0-9\.]*\w+\.\w+(?= differ))|((Only in \w+\: )\K(.*))' | while read l; do check_filename $l; done
```
Will need to download the bucket metadata first before this command runs.

lkatalin replied:
Do you have a reference or example for this? Is it like https://github.com/sigstore/root-signing/blob/main/.github/workflows/sync-preprod-to-prod.yml#L89?

Hayden-IO replied:
Correct!

lkatalin replied:
Thanks, it makes sense now that I've looked at it again.

Hayden-IO commented on .github/workflows/sync-main-to-prod.yml:
```diff
@@ -77,20 +100,21 @@ jobs:
             fi;
 
-            gcloud --quiet storage cp --cache-control=no-store -r repository/repository/$f gs://sigstore-preprod-tuf-root/
+            gcloud --quiet storage cp --cache-control=no-store -r repository/repository/$f gs://sigstore-tuf-root/
```
Per @kommendorkapten's suggestion in https://github.com/sigstore/root-signing/issues/893#issuecomment-1677180668, can we keep syncing to preprod and also sync to prod? This'll keep the probers happy

lkatalin replied:
Updated

lkatalin committed:
Update sync-main-to-preprod.yml to sync also to prod

This sync will now occur on the condition that
only timestamp or snapshot files are changed in main.

This change renames sync-main-to-preprod.yml to sync-main-to-preprod-and-prod.yml.

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin committed:
Update sync-ceremony-to-main.yml to sync also to preprod

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin committed:
Update orchestration.md to reference manual sync to prod

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin commented:
I think this is good to go, but just trying to think through to make sure there aren't any unfortunate race conditions between the ceremony syncs and the timestamp syncs. I *think* we avoid those by only [diffing with prod](https://github.com/sigstore/root-signing/pull/916/files#diff-74730d591b8b3922955b8495b127e391af70b811608729df3d600bd909447854R98-R100) and not preprod in the timestamp sync, since both flows will be updating preprod.

Hayden-IO commented on playbooks/ORCHESTRATION.md:
```diff
@@ -242,13 +242,15 @@ Submitting this PR will trigger a push to the preproduction GCS bucket, so ensur
 ## Post-ceremony Steps
 
-1. If any root keyholders have changed, update the [current root keyholders](https://github.com/sigstore/root-signing#current-sigstore-root-keyholders) with their name, key ID, and location of their key material.
+1. The preproduction GCS bucket will need to be manually synced to the GCS production bucket as of [916](https://github.com/sigstore/root-signing/pull/916).
```
We can do this as a follow up, but we should link to the GHA (which doesn't yet exist)

Hayden-IO commented:
@lkatalin Good callout about race conditions. Also note that the timestamp/snapshot GHA should not run while a ceremony is ongoing (https://github.com/sigstore/root-signing/blob/main/.github/workflows/stable-snapshot-timestamp.yml#L32-L42), so that should also help to mitigate an overlap.

Hayden-IO commented:
@bobcallaway @kommendorkapten I'd appreciate another set of eyes just to make sure this looks good.

bobcallaway commented on .github/workflows/sync-ceremony-to-main.yml:
```diff
@@ -63,3 +62,82 @@ jobs:
           pr_title: "Merge ceremony branch ${{ inputs.branch || github.ref_name }} into ${{ github.event.repository.default_branch }}"
           pr_body: "Merge ceremony branch to main"
           pr_reviewer: asraa,dlorenc,haydentherapper,joshuagl,kommendorkapten
```
should this reviewer list be a team that we can define via pulumi vs having this hard-coded?

Hayden-IO replied:
We could also omit it entirely, as automation should handle reviewing.

lkatalin replied:
I agree, though since that was there before, I'd normally suggest a different PR for it. But I can also just remove it in a separate commit if that seems better.

Hayden-IO replied:
Id say to leave it for a later change.

kommendorkapten reviewed (approved):
🚀

Hayden-IO commented:
Thank you @lkatalin! I'll keep an eye on GHAs tomorrow when the new timestamp/snapshot is generated.
